execution-env.md

```diff
@@ -1,80 +1,119 @@
+
+* Be discarded for writes
+* Access any location within the resource for reads and writes
+* Return zero values for reads or (0, 0, 0, X) with X being 0, 1, -1, or extrema for integers, or -0.0, +0.0, -1.0, +1.0 for floating point values
+* Atomics can return undefined values.
 
+The robust resource access behavior of OpenGL is extended to allow for trapping behavior as well:
+
+* Cause a trap that immediately stops execution of the shader, and fills all of the shader stage's output with zeroes.
+
+Additionnally functionality exists to create pointers to values inside resources.
+Creating a pointer for a resource that points to outside the resource could also result in the following behavior:
+
+* Cause a trap.
+* Defer the out of bounds behavior to when the pointer is dereferenced.
+* Create a pointer to any location within the resource. (this is equivalent to deferring and "access any location" for resource accesses).
+
+(TODO are there pointers to unsized arrays? Could talk about the "sized part of the type if that's the case")
+
+## WebGPU Execution Environment Specification for SPIR-V
 
 ### Introduction
 
 WebGPU is a client API for SPIR-V. Valid s<span style="color:#222222;">haders for WebGPU are defined by this document in addition to the [Khronos SPIR-V Specification, version 1.3, revision 5](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html)<span style="color:#222222;">.<span style="color:#222222;">  [TBD: update to latest available specification] See the "Version" subsection below for allowed version numbers.</span></span></span>
 
 The following sections visit the parts of a SPIR-V module, in order, covering each section, and state additional requirements for using shaders in WebGPU. These are in addition to all requirements already present in the SPIR-V specification itself.
 
-
 ### Undefined Behaviour
 
 The SPIR-V specification states that certain actions by a SPIR-V shader result in _undefined_ _behaviour_.  A WebGPU implementation is required to be secure.  Therefore in the context of WebGPU, a shader which triggers SPIR-V's undefined behaviour may continue to execute in an arbitrary manner except that it may not sense or mutate any state outside its ExecutionContext, and it may not cause or sense I/O outside its ExecutionContext.
 
 NOTE: One of the possible behaviours of an ill-behaved application is non-termination.  The WebGPU specification should say what happens to a runaway application. TBD
 
+### Out of bounds
+
+Operations that are considered as out of bounds in SPIR-V (that are valid for WebGPU) are the following:
+
+* `OpLoad`, `OpRead`, `OpMemoryCopy` when the pointer is out of bounds.
+* `OpAccessChain` when any of the indices for array accesses are larger (or equal) to the array's size or when the pointer is out of bounds.
+  Note that the size of arrays might known at runtime for `OpTypeRuntimeArray`.
```

Comment:
In the statement about OpTypeRuntimeArray, change "might" to "are".

Should also say that when the pipeline is created, the final values are known for specialization constants.  Therefore the sizes of all OpTypeArray instances are known.

Reply:
Done